File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/filter/DeleteTest.java
package org.babyfish.jimmer.sql.filter;

import org.babyfish.jimmer.sql.ast.mutation.AffectedTable;
import org.babyfish.jimmer.sql.common.AbstractMutationTest;
import org.babyfish.jimmer.sql.filter.common.FileFilter;
import org.babyfish.jimmer.sql.model.filter.File;
import org.babyfish.jimmer.sql.model.filter.FileProps;
import org.junit.jupiter.api.Test;

public class DeleteTest extends AbstractMutationTest {

    @Test
    public void test() {
        FileFilter.withUser(2L, () -> {
            executeAndExpectResult(
                    getSqlClient(it -> {
                        it.addFilters(new FileFilter());
                        it.setMaxCommandJoinCount(0);
                    }).getEntities().deleteCommand(File.class, 8L),
                    ctx -> {
                        ctx.statement(it -> {
                            it.sql("delete from FILE_USER_MAPPING where FILE_ID = ?");
                            it.variables(8L);
                        });
                        ctx.statement(it -> {
                            it.sql("select tb_1_.ID from FILE tb_1_ where tb_1_.PARENT_ID = ?");
                            it.variables(8L);
                        });
                        ctx.statement(it -> {
                            it.sql(
                                    "select tb_1_.ID from FILE tb_1_ inner join FILE tb_2_ on tb_1_.PARENT_ID = tb_2_.ID where tb_2_.ID in (?, ?, ?, ?, ?)"
                            );
                            it.variables(9L, 10L, 11L, 12L, 13L);
                        });
                        ctx.statement(it -> {
                            it.sql(
                                    "select FILE_ID, USER_ID " +
                                            "from FILE_USER_MAPPING " +
                                            "where FILE_ID in (?, ?, ?, ?, ?)"
                            );
                            it.variables(9L, 10L, 11L, 12L, 13L);
                        });
                        ctx.statement(it -> {
                            it.sql(
                                    "delete from FILE_USER_MAPPING " +
                                            "where FILE_ID = ? and USER_ID = ?"
                            );
                            it.batchVariables(0, 9L, 2L);
                            it.batchVariables(1, 9L, 3L);
                            it.batchVariables(2, 10L, 3L);
                            it.batchVariables(3, 10L, 4L);
                            it.batchVariables(4, 11L, 2L);
                            it.batchVariables(5, 11L, 4L);
                            it.batchVariables(6, 12L, 2L);
                            it.batchVariables(7, 12L, 3L);
                            it.batchVariables(8, 13L, 3L);
                            it.batchVariables(9, 13L, 4L);
                        });
                        ctx.statement(it -> {
                            it.sql("delete from FILE where ID in (?, ?, ?, ?, ?)");
                            it.variables(9L, 10L, 11L, 12L, 13L);
                        });
                        ctx.statement(it -> {
                            it.sql("delete from FILE where ID = ?");
                            it.variables(8L);
                        });
                        ctx.rowCount(AffectedTable.of(File.class), 6);
                        ctx.rowCount(AffectedTable.of(FileProps.USERS), 13);
                    }
            );
        });
    }
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/filter/DeleteTest.kt
package org.babyfish.jimmer.sql.kt.filter

import org.babyfish.jimmer.sql.ast.mutation.QueryReason
import org.babyfish.jimmer.sql.kt.common.AbstractMutationTest
import org.babyfish.jimmer.sql.kt.filter.common.FileFilter
import org.babyfish.jimmer.sql.kt.model.filter.File
import kotlin.test.Test

class DeleteTest : AbstractMutationTest() {

    private val _sqlClient = sqlClient {
        addFilters(FileFilter())
    }

    @Test
    fun test() {
        FileFilter.withUser(2L) {
            executeAndExpectResult({
                _sqlClient.entities.delete(File::class, 8L, it)
            }) {
                statement {
                    sql("select USER_ID from FILE_USER_MAPPING where FILE_ID = ?")
                    variables(8L)
                    queryReason(QueryReason.UPSERT_NOT_SUPPORTED)
                }
                statement {
                    sql("delete from FILE_USER_MAPPING where FILE_ID = ? and USER_ID = ?")
                    batchVariables(0, 8L, 2L)
                    batchVariables(1, 8L, 3L)
                }
                statement {
                    sql(
                        """select tb_1_.ID 
                            |from FILE tb_1_ 
                            |inner join FILE tb_2_ on tb_1_.PARENT_ID = tb_2_.ID 
                            |inner join FILE tb_3_ on tb_2_.PARENT_ID = tb_3_.ID 
                            |where tb_3_.PARENT_ID = ?""".trimMargin()
                    )
                    variables(8L)
                    queryReason(QueryReason.TOO_DEEP)
                }
                statement {
                    sql(
                        """select tb_1_.FILE_ID, tb_1_.USER_ID 
                            |from FILE_USER_MAPPING tb_1_ 
                            |inner join FILE tb_2_ on tb_1_.FILE_ID = tb_2_.ID 
                            |where exists(
                            |--->select * 
                            |--->from FILE tb_3_ 
                            |--->where tb_2_.PARENT_ID = tb_3_.ID and tb_3_.PARENT_ID = ?
                            |)""".trimMargin()
                    )
                    variables(8L)
                }
                statement {
                    sql(
                        """delete from FILE tb_1_ 
                            |where exists(
                            |--->select * 
                            |--->from FILE tb_2_ 
                            |--->where tb_1_.PARENT_ID = tb_2_.ID and tb_2_.PARENT_ID = ?
                            |)""".trimMargin()
                    )
                    variables(8L)
                }
                statement {
                    sql(
                        "select tb_1_.FILE_ID, tb_1_.USER_ID " +
                            "from FILE_USER_MAPPING tb_1_ " +
                            "inner join FILE tb_2_ on tb_1_.FILE_ID = tb_2_.ID " +
                            "where tb_2_.PARENT_ID = ?"
                    )
                    variables(8L)
                }
                statement {
                    sql(
                        """delete from FILE_USER_MAPPING where FILE_ID = ? and USER_ID = ?""".trimMargin()
                    )
                    batchVariables(0, 9L, 2L)
                    batchVariables(1, 9L, 3L)
                    batchVariables(2, 10L, 3L)
                    batchVariables(3, 10L, 4L)
                    batchVariables(4, 11L, 2L)
                    batchVariables(5, 11L, 4L)
                    batchVariables(6, 12L, 2L)
                    batchVariables(7, 12L, 3L)
                    batchVariables(8, 13L, 3L)
                    batchVariables(9, 13L, 4L)
                }
                statement {
                    sql("delete from FILE where PARENT_ID = ?")
                    variables(8L)
                }
                statement {
                    sql("delete from FILE where ID = ?")
                    variables(8L)
                }
                rowCount(File::class, 6)
                rowCount(File::users, 13)
            }
        }
    }
}